toolchain/check/handle_named_constraint.cpp

```diff
@@ -3,29 +3,176 @@
+        context.named_constraints().Add(constraint_info);
+    if (constraint_info.has_parameters()) {
+      constraint_decl.type_id = GetGenericNamedConstraintType(
+          context, constraint_decl.named_constraint_id,
+          context.scope_stack().PeekSpecificId());
+    }
+  }
+
+  // Write the completed NamedConstraintDecl instruction.
+  ReplaceInstBeforeConstantUse(context, decl_inst_id, constraint_decl);
+
+  (void)is_definition;
+  return {constraint_decl.named_constraint_id, decl_inst_id};
+}
+
+auto HandleParseNode(Context& context, Parse::NamedConstraintDeclId node_id)
+    -> bool {
+  BuildNamedConstraintDecl(context, node_id, /*is_definition=*/false);
+  context.decl_name_stack().PopScope();
+  return true;
 }
 
 auto HandleParseNode(Context& context,
                      Parse::NamedConstraintDefinitionStartId node_id) -> bool {
-  return context.TODO(node_id, "HandleNamedConstraintDefinitionStart");
+  auto [named_constraint_id, decl_inst_id] =
+      BuildNamedConstraintDecl(context, node_id, /*is_definition=*/true);
+  auto& constraint_info = context.named_constraints().Get(named_constraint_id);
+
+  // TODO: Support for `template constraint`.
+  bool is_template = false;
+
+  // Track that this declaration is the definition.
+  CARBON_CHECK(!constraint_info.has_definition_started(),
+               "Can't merge with defined named constraints.");
+  constraint_info.definition_id = decl_inst_id;
+  constraint_info.scope_id = context.name_scopes().Add(
+      decl_inst_id, SemIR::NameId::None, constraint_info.parent_scope_id);
+
+  auto self_specific_id =
+      context.generics().GetSelfSpecific(constraint_info.generic_id);
+  StartGenericDefinition(context, constraint_info.generic_id);
+
+  context.inst_block_stack().Push();
+
+  // Declare and introduce `Self`. We model `Self` as a symbolic binding whose
+  // type is the named constraint.
+  SemIR::TypeId self_type_id =
+      GetNamedConstraintType(context, named_constraint_id, self_specific_id);
+  constraint_info.self_param_id = GetSelfParameter(
+      context, self_type_id, constraint_info.scope_id, is_template);
+
+  // Enter the constraint scope.
+  context.scope_stack().PushForEntity(decl_inst_id, constraint_info.scope_id,
+                                      self_specific_id);
+
+  constraint_info.body_block_id = context.inst_block_stack().PeekOrAdd();
+
+  context.node_stack().Push(node_id, named_constraint_id);
+  return true;
 }
 
 auto HandleParseNode(Context& context,
-                     Parse::NamedConstraintIntroducerId node_id) -> bool {
-  return context.TODO(node_id, "HandleNamedConstraintIntroducer");
+                     Parse::NamedConstraintDefinitionId /*node_id*/) -> bool {
+  auto named_constraint_id =
+      context.node_stack()
+          .Pop<Parse::NodeKind::NamedConstraintDefinitionStart>();
+  context.inst_block_stack().Pop();
+
+  auto& constraint_info = context.named_constraints().Get(named_constraint_id);
```

Comment:
```suggestion
  const auto& constraint_info = context.named_constraints().Get(named_constraint_id);
```

nit: does this need to be non-const? I don't normally suggest it, but with the reference it feels like it may be useful

Reply:
Done